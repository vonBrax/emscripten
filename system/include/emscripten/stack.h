/*
 * Copyright 2020 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#pragma once

// API that gives access to introspecting the Wasm data stack. Build with
// -lstack.js to use this API.

#ifdef __cplusplus
extern "C" {
#endif

// Returns the starting address of the wasm stack. This is the address
// that the stack pointer would point to when no bytes are in use on the stack.
uintptr_t emscripten_stack_get_base(void);

// Returns the end address of the wasm stack. This is the address that the stack
// pointer would point to when the whole stack is in use.  (the address pointed
// to by the end is not part of the stack itself) Note that in fastcomp, the
// stack grows up, whereas in wasm backend, it grows down.  So with wasm
// backend, the address returned by emscripten_stack_get_end() is smaller than
// emscripten_stack_get_base().
uintptr_t emscripten_stack_get_end(void);

// Returns the current stack pointer.
uintptr_t emscripten_stack_get_current(void);

// Returns the number of free bytes left on the stack.
size_t emscripten_stack_get_free(void);

#ifdef __cplusplus
}
#endif
