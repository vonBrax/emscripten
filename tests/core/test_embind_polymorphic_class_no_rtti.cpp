// Copyright 2020 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.

#include <emscripten.h>
#include <emscripten/bind.h>
#include <emscripten/val.h>
#include <stdio.h>

EM_JS(void, calltest, (), {
  var foo = new Module.Foo();
  console.log("foo.test() returned: " + foo.test());
});

class Foo {
  public:
  virtual ~Foo() = default;
  virtual int test() = 0;
};

class Bar : public Foo {
  public:
  int test() override { return 42; }
};

int main(int argc, char** argv){
  printf("Hello, world.\n");
  calltest();
  return 0;
}

std::shared_ptr<Foo> MakeFoo() {
  return std::make_shared<Bar>();
}

EMSCRIPTEN_BINDINGS(my_module) {
  emscripten::class_<Foo>("Foo")
    .smart_ptr_constructor("Foo", &MakeFoo)
    .function("test", &Foo::test)
    ;
};
