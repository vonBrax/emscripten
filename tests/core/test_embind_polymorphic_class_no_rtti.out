Hello, world.
foo.test() returned: 42
